File: crates/fey_math/src/vec4.rs
use crate::{Num, Vec2, Vec3, impl_vec, vec2, vec3};
use std::fmt::{Display, Formatter};

pub type Vec4F = Vec4<f32>;
pub type Vec4I = Vec4<i32>;
pub type Vec4U = Vec4<u32>;

/// A 4-dimensional vector.
#[repr(C)]
#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Vec4<T> {
    pub x: T,
    pub y: T,
    pub z: T,
    pub w: T,
}

/// Create a [`Vec4`].
#[inline]
pub const fn vec4<T>(x: T, y: T, z: T, w: T) -> Vec4<T> {
    Vec4 { x, y, z, w }
}

impl_vec!(
    NAME = Vec4
    SHORT = vec4
    LEN = 4
    FIELDS = (x, y, z, w)
    TUPLE = (T, T, T, T)
);

impl<T> Vec4<T> {
    /// Return the vector with the x-value replaced.
    #[inline]
    pub fn with_x(self, x: T) -> Self {
        Self { x, ..self }
    }

    /// Return the vector with the y-value replaced.
    #[inline]
    pub fn with_y(self, y: T) -> Self {
        Self { y, ..self }
    }

    /// Return the vector with the z-value replaced.
    #[inline]
    pub fn with_z(self, z: T) -> Self {
        Self { z, ..self }
    }

    /// Return the vector with the w-value replaced.
    #[inline]
    pub fn with_w(self, w: T) -> Self {
        Self { w, ..self }
    }
}

impl<T: Copy> Vec4<T> {
    /// Reduce to a 2D vector by dropping `z` and `w`.
    #[inline]
    pub const fn xy(self) -> Vec2<T> {
        vec2(self.x, self.y)
    }

    /// Reduce to a 3D vector by dropping `w`.
    #[inline]
    pub const fn xyz(self) -> Vec3<T> {
        vec3(self.x, self.y, self.z)
    }

    /// Swizzle components.
    #[inline]
    pub fn yxzw(self) -> Self {
        vec4(self.y, self.x, self.z, self.w)
    }

    /// Swizzle components.
    #[inline]
    pub fn zxyw(self) -> Self {
        vec4(self.z, self.x, self.y, self.w)
    }

    /// Swizzle components.
    #[inline]
    pub fn wxzy(self) -> Self {
        vec4(self.w, self.x, self.z, self.y)
    }

    /// Swizzle components.
    #[inline]
    pub fn yxwz(self) -> Self {
        vec4(self.y, self.x, self.w, self.z)
    }

    /// Swizzle components.
    #[inline]
    pub fn yzxw(self) -> Self {
        vec4(self.y, self.z, self.x, self.w)
    }

    /// Swizzle components.
    #[inline]
    pub fn zyxw(self) -> Self {
        vec4(self.z, self.y, self.x, self.w)
    }

    /// Swizzle components.
    #[inline]
    pub fn wzxy(self) -> Self {
        vec4(self.w, self.z, self.x, self.y)
    }

    /// Swizzle components.
    #[inline]
    pub fn yzwx(self) -> Self {
        vec4(self.y, self.z, self.w, self.x)
    }

    /// Swizzle components.
    #[inline]
    pub fn zywx(self) -> Self {
        vec4(self.z, self.y, self.w, self.x)
    }

    /// Swizzle components.
    #[inline]
    pub fn wzyx(self) -> Self {
        vec4(self.w, self.z, self.y, self.x)
    }

    /// Swizzle components.
    #[inline]
    pub fn ywzx(self) -> Self {
        vec4(self.y, self.w, self.z, self.x)
    }
}

impl<T: Num> Vec4<T> {
    /// A unit vector representing the x-axis, equal to `(1, 0, 0, 0)`.
    pub const X_AXIS: Self = vec4(T::ONE, T::ZERO, T::ZERO, T::ZERO);

    /// A unit vector representing the y-axis, equal to `(0, 1, 0, 0)`.
    pub const Y_AXIS: Self = vec4(T::ZERO, T::ONE, T::ZERO, T::ZERO);

    /// A unit vector representing the z-axis, equal to `(0, 0, 1, 0)`.
    pub const Z_AXIS: Self = vec4(T::ZERO, T::ZERO, T::ONE, T::ZERO);

    /// A unit vector representing the w-axis, equal to `(0, 0, 0, 1)`.
    pub const W_AXIS: Self = vec4(T::ZERO, T::ZERO, T::ZERO, T::ONE);
}

impl<T: Display> Display for Vec4<T> {
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        self.x.fmt(f)?;
        f.write_str(", ")?;
        self.y.fmt(f)?;
        f.write_str(", ")?;
        self.z.fmt(f)?;
        f.write_str(", ")?;
        self.w.fmt(f)
    }
}

impl<T: Num> From<Vec2<T>> for Vec4<T> {
    #[inline]
    fn from(Vec2 { x, y }: Vec2<T>) -> Self {
        vec4(x, y, T::ZERO, T::ZERO)
    }
}

impl<T: Num> From<Vec3<T>> for Vec4<T> {
    #[inline]
    fn from(Vec3 { x, y, z }: Vec3<T>) -> Self {
        vec4(x, y, z, T::ZERO)
    }
}

File: crates/fey_math/lua/Vec4.lua
---@meta

---A 4-dimensional vector.
---@class (exact) Vec4: Vec4Methods
---@field x number
---@field y number
---@field z number
---@field w number
---@operator add(Vec4): Vec4
---@operator sub(Vec4): Vec4
---@operator mul(Vec4|number): Vec4
---@operator div(Vec4|number): Vec4
---@operator unm: Vec4

---@class Vec4Class : Vec4Methods
---@overload fun(x: number, y: number, z: number, w: number): Vec4
local module = {}

---@class Vec4Methods
local methods = {}

---`(0, 0, 0, 0)`
---@return Vec4
---@nodiscard
function module.zero() end

---`(1, 1, 1, 1)`
---@return Vec4
---@nodiscard
function module.one() end

---`(1, 0, 0, 0)`
---@return Vec4
---@nodiscard
function module.x_axis() end

---`(0, 1, 0, 0)`
---@return Vec4
---@nodiscard
function module.y_axis() end

---`(0, 0, 1, 0)`
---@return Vec4
---@nodiscard
function module.z_axis() end

---`(0, 0, 0, 1)`
---@return Vec4
---@nodiscard
function module.w_axis() end

---Create a new vector.
---@param x number
---@param y number
---@param z number
---@param w number
---@return Vec4
---@nodiscard
function module.new(x, y, z, w) end

---Create a new vector `(val, val, val, val)`.
---@param val number
---@return Vec4
---@nodiscard
function module.splat(val) end

---Returns a temporary copy of this value.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.clone(self) end

---Boxes the value if not already boxed.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.box(self) end

---Boxes a copy of this value.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.box_clone(self) end

---Returns a copy of this vector with absolute value of all components.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.abs(self) end

---Returns `true` if the two vectors are approximately equal.
---@param self Vec4
---@param other Vec4
---@return boolean
---@nodiscard
function methods.approx(self, other) end

---Returns `true` if the vector is approximately equal to `(0, 0, 0)`.
---@param self Vec4
---@return boolean
---@nodiscard
function methods.approx_zero(self) end

---Rounds the vector's components up to the nearest integer.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.ceil(self) end

---Clamp's the vector's components between the components of `min` and `max`.
---@param self Vec4
---@param min Vec4
---@param max Vec4
---@return Vec4
---@nodiscard
function methods.clamp(self, min, max) end

---Returns the normalized direction towards the point.
---@param self Vec4
---@param point Vec4
---@return Vec4
---@nodiscard
function methods.dir_to(self, point) end

---Returns the distance between two points.
---@param self Vec4
---@param point Vec4
---@return Vec4
---@nodiscard
function methods.dist(self, point) end

---Returns the dot product of two vectors.
---@param self Vec4
---@param other Vec4
---@return number
---@nodiscard
function methods.dot(self, other) end

---Rounds the vector's components down to the nearest integer.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.floor(self) end

---Returns `true` if the vector is equal to `(0, 0, 0)`.
---@return boolean
---@nodiscard
function methods.is_zero(self) end

---Length of the vector.
---@param self Vec4
---@return number
---@nodiscard
function methods.len(self) end

---Returns a vector with components set to the largest values in the arguments.
---@param self Vec4
---@param ... Vec4
---@return Vec4
---@nodiscard
function methods.max(self, ...) end

---Returns a vector with components set to the smallest values in the arguments.
---@param self Vec4
---@param ... Vec4
---@return Vec4
---@nodiscard
function methods.min(self, ...) end

---Normalizes the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.norm(self) end

---Reflects the vector off the provided normal.
---@param self Vec4
---@param normal Vec4
---@return Vec4
---@nodiscard
function methods.reflect(self, normal) end

---Rounds the vector's components to the nearest integer.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.round(self) end

---Returns a copy of this vector with all components signed (set to `1`, `-1`, or `0`).
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.sign(self) end

---Returns the square between two points.
---@param self Vec4
---@param point Vec4
---@return Vec4
---@nodiscard
function methods.sqr_dist(self, point) end

---Squared length of the vector.
---@param self Vec4
---@return number
---@nodiscard
function methods.sqr_len(self) end

---Changes the length of the vector without changing its direction.
---@param self Vec4
---@param new_len number
---@return number
---@nodiscard
function methods.with_len(self, new_len) end

---Replaces the `x` component of the vector.
---@param self Vec4
---@param x number
---@return Vec4
---@nodiscard
function methods.with_x(self, x) end

---Replaces the `y` component of the vector.
---@param self Vec4
---@param y number
---@return Vec4
---@nodiscard
function methods.with_y(self, y) end

---Replaces the `z` component of the vector.
---@param self Vec4
---@param z number
---@return Vec4
---@nodiscard
function methods.with_z(self, z) end

---Replaces the `w` component of the vector.
---@param self Vec4
---@param w number
---@return Vec4
---@nodiscard
function methods.with_w(self, w) end

---Returns the 2D portion of this vector.
---@param self Vec4
---@return Vec2
---@nodiscard
function methods.xy(self) end

---Returns the 3D portion of this vector.
---@param self Vec4
---@return Vec3
---@nodiscard
function methods.xyz(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.wxzy(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.wzxy(self) end



---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.wzyx(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.ywzx(self) end



---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.yxwz(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.yxzw(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.yzwx(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.yzxw(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.zxyw(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.zywx(self) end

---Swizzles the vector.
---@param self Vec4
---@return Vec4
---@nodiscard
function methods.zyxw(self) end

return module
